rfcs/45-stageless.md

```diff
@@ -0,0 +1,545 @@
+
+- `A` is strictly before `B`
+- the data locks (in terms of read and write accesses to the `World`) of system `A` will not be released until system `B` has completed
+- the data locks from system `A` are ignored for the purposes of checking if system `B` can run
+  - after all, they're not actualy locked, merely reserved
+  - competing read-locks caused by other systems will still need to complete before a data store can be mutated; the effect is to "ignore the locks of `A`", not "skip the check completely"
+
+In addition to their use in run criteria, atomic ordering constraints are extremely useful for forcing tight interlocking behavior between systems.
+They allow us to ensure that the state read from and written to the earlier system cannot be invalidated.
+This functionality is extremely useful when updating indexes (used to look up components by value):
+allowing you to ensure that the index is still valid when the system using the index uses it.
+
+Of course, atomic ordering constraints should be used thoughtfully.
+As the strictest of the three types of system ordering dependency, they can easily result in unsatisfiable schedules if applied to large groups of systems at once.
+
+On the other hand, atomic ordering constraints can be helpful when attempting to split large complex systems into multiple parts.
+By guaranteeing that the initial state of your complex system cannot be altered before the later parts of the system are complete,
+you can safely parallelize the rest of the work into seperate systems, improving both performance and maintainability.
+
+### States
+
+**States** are one particularly common and powerful form of run criteria, allowing you to toggle systems on-and-off based on the value of a given resource and smoothly handle transitions between states by running cleanup and initialization systems.
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three sets of systems:
+
+1. **Update systems:** these systems run each frame if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(apply_damage.in_state(GameState::Playing))`
+2. **On-enter systems:** these systems run once when the specified state is entered.
+   1. `app.add_system(generate_map.on_enter(GameState::Playing))`
+3. **On-exit systems:** these systems run once when the specified stated is exited.
+   1. `app.add_system(autosave.on_exit(GameState::Playing))`
+
+Update systems are by far the simplest: they're simply powered by run criteria.
+`.in_state` is precisely equivalent to `run_if_resource_is`, except with an additional trait bound that the resource must implement `State`.

```

Comment:
This looks like the first mention of this convenience function. It isn't described before this.